skafka/src/main/scala/com/evolutiongaming/skafka/consumer/RebalanceCallback.scala

```diff
@@ -0,0 +1,267 @@
+    } yield result
+
+  def offsetsForTimes[F[_]: MonadThrowable](
+    timestampsToSearch: Nem[TopicPartition, Instant]
+  ): RebalanceCallback[F, Map[TopicPartition, Option[OffsetAndTimestamp]]] =
+    for {
+      result <- ConsumerStep(_.offsetsForTimes(timestampsToSearchJ(timestampsToSearch)))
+      result <- lift(offsetsAndTimestampsMapF[F](result))
+    } yield result
+
+  def offsetsForTimes[F[_]: MonadThrowable](
+    timestampsToSearch: Nem[TopicPartition, Instant],
+    timeout: FiniteDuration
+  ): RebalanceCallback[F, Map[TopicPartition, Option[OffsetAndTimestamp]]] =
+    for {
+      result <- ConsumerStep(_.offsetsForTimes(timestampsToSearchJ(timestampsToSearch), timeout.asJava))
+      result <- lift(offsetsAndTimestampsMapF[F](result))
+    } yield result
+
+  def partitionsFor[F[_]: ApplicativeThrowable](topic: Topic): RebalanceCallback[F, List[PartitionInfo]] =
+    for {
+      result <- ConsumerStep(_.partitionsFor(topic))
+      result <- lift(partitionsInfoListF[F](result))
+    } yield result
+
+  def partitionsFor[F[_]: ApplicativeThrowable](
+    topic: Topic,
+    timeout: FiniteDuration
+  ): RebalanceCallback[F, List[PartitionInfo]] =
+    for {
+      result <- ConsumerStep(_.partitionsFor(topic, timeout.asJava))
+      result <- lift(partitionsInfoListF[F](result))
+    } yield result
+
+  def paused[F[_]]: RebalanceCallback[F, Set[TopicPartition]] =
+    ConsumerStep(
+      _.paused().asScala.toSet.map(unsafeTopicPartition)
+    )
+
+  // TODO using Offset.unsafe to avoid requiring ApplicativeThrowable as in Offset.of
+  //  also it looks like consumer.position cannot return invalid offset
+  def position[F[_]](tp: TopicPartition): RebalanceCallback[F, Offset] =
+    ConsumerStep(c => Offset.unsafe(c.position(tp.asJava)))
+
+  def position[F[_]](tp: TopicPartition, timeout: FiniteDuration): RebalanceCallback[F, Offset] =
+    ConsumerStep(c => Offset.unsafe(c.position(tp.asJava, timeout.asJava)))
+
+  def seek[F[_]](partition: TopicPartition, offset: Offset): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seek(partition.asJava, offset.value))
+
+  def seek[F[_]](partition: TopicPartition, offsetAndMetadata: OffsetAndMetadata): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seek(partition.asJava, offsetAndMetadata.asJava))
+
+  def seekToBeginning[F[_]](partitions: Nes[TopicPartition]): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seekToBeginning(partitions.asJava))
+
+  def seekToEnd[F[_]](partitions: Nes[TopicPartition]): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seekToEnd(partitions.asJava))
+
+  def subscription[F[_]]: RebalanceCallback[F, Set[Topic]] = ConsumerStep(_.subscription().asScala.toSet)
+
+  private[consumer] def run[F[_]: Monad: ToTry, A](
+    rc: RebalanceCallback[F, A],
+    consumer: ConsumerJ[_, _]
+  ): Try[A] = {
+
+    // TODO @tailrec failed to convert map/flatMap to patmat in 5 minutes, was getting scalac errors about type mismatches
```

Comment:
you would need to get rid of tailing `map` & `flatMap` calls

Reply:
changed in 3d26df19caf2b34ca4cc429e920c98c5c77e5655